codec/grpc/grpc_codec.go

```diff
@@ -1,33 +1,63 @@
 package grpc
 
-import "fmt"
+import (
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/mem"
+
+	// Guarantee that the built-in proto is called registered before this one
+	// so that it can be replaced.
+	_ "google.golang.org/grpc/encoding/proto"
+)
 
 // Name is the name registered for the proto compressor.
 const Name = "proto"
 
-type Codec struct{}
-
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToSizedBufferVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
+}
+
+type Codec struct {
+	fallback encoding.CodecV2
 }
 
-func (Codec) Marshal(v interface{}) ([]byte, error) {
-	vt, ok := v.(vtprotoMessage)
-	if !ok {
-		return nil, fmt.Errorf("failed to marshal, message is %T (missing vtprotobuf helpers)", v)
+func (Codec) Name() string { return Name }
+
+func (c *Codec) Marshal(v any) (mem.BufferSlice, error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
+			if _, err := m.MarshalToSizedBufferVT(buf[:size]); err != nil {
+				return nil, err
+			}
+			return mem.BufferSlice{mem.SliceBuffer(buf)}, nil
+		}
+		pool := mem.DefaultBufferPool()
+		buf := pool.Get(size)
+		if _, err := m.MarshalToSizedBufferVT((*buf)[:size]); err != nil {
+			pool.Put(buf)
+			return nil, err
+		}
+		return mem.BufferSlice{mem.NewBuffer(buf, pool)}, nil
 	}
-	return vt.MarshalVT()
+
+	return c.fallback.Marshal(v)
 }
 
-func (Codec) Unmarshal(data []byte, v interface{}) error {
-	vt, ok := v.(vtprotoMessage)
-	if !ok {
-		return fmt.Errorf("failed to unmarshal, message is %T (missing vtprotobuf helpers)", v)
+func (c *Codec) Unmarshal(data mem.BufferSlice, v any) error {
+	if m, ok := v.(vtprotoMessage); ok {
+		buf := data.MaterializeToBuffer(mem.DefaultBufferPool())
+		defer buf.Free()
+		return m.UnmarshalVT(buf.ReadOnlyData())
 	}
-	return vt.UnmarshalVT(data)
+
+	return c.fallback.Unmarshal(data, v)
 }
 
-func (Codec) Name() string {
-	return Name
+func init() {
+	encoding.RegisterCodecV2(&Codec{
+		fallback: encoding.GetCodecV2("proto"),
```

Comment:
We can use name from proto so we will avoid unused import https://pkg.go.dev/google.golang.org/grpc/encoding/proto#pkg-constants
```suggestion
		fallback: encoding.GetCodecV2(proto.Name),
```